File: apps/desktop/src/api/automation.ts
// Updated Nov 16, 2025: Added comprehensive error handling, input validation, and timeout handling
import { invoke } from '@tauri-apps/api/core';
import type { CaptureResult } from '../types/capture';
import type {
  AutomationClickRequest,
  AutomationElementInfo,
  AutomationOcrResult,
  AutomationQuery,
  AutomationScreenshotOptions,
  OverlayClickPayload,
  OverlayRegionPayload,
  OverlayTypePayload,
} from '../types/automation';
import { normalizeCaptureResult, type RawCaptureResult } from '../utils/captureTransforms';

// Updated Nov 16, 2025: Configurable timeout for automation operations
const AUTOMATION_TIMEOUT_MS = 30000; // 30 seconds default

// Updated Nov 16, 2025: Wrapper for invoke with timeout and error handling
async function invokeWithTimeout<T>(
  command: string,
  args?: Record<string, unknown>,
  timeoutMs: number = AUTOMATION_TIMEOUT_MS,
): Promise<T> {
  return new Promise((resolve, reject) => {
    const timeoutId = setTimeout(() => {
      reject(new Error(`Automation command '${command}' timed out after ${timeoutMs}ms`));
    }, timeoutMs);

    invoke<T>(command, args)
      .then((result) => {
        clearTimeout(timeoutId);
        resolve(result);
      })
      .catch((error) => {
        clearTimeout(timeoutId);
        reject(new Error(`Automation command '${command}' failed: ${error}`));
      });
  });
}

// Updated Nov 16, 2025: Input validation helper
function validateNonEmpty(value: string | undefined, fieldName: string): void {
  if (!value || value.trim().length === 0) {
    throw new Error(`${fieldName} cannot be empty`);
  }
}

// Updated Nov 16, 2025: Validate coordinates
function validateCoordinates(x: number | undefined, y: number | undefined): void {
  if (x !== undefined && (x < 0 || !Number.isFinite(x))) {
    throw new Error(`Invalid x coordinate: ${x}`);
  }
  if (y !== undefined && (y < 0 || !Number.isFinite(y))) {
    throw new Error(`Invalid y coordinate: ${y}`);
  }
}

interface RawBoundingRect {
  left: number;
  top: number;
  width: number;
  height: number;
}

interface RawAutomationElementInfo {
  id: string;
  name: string;
  class_name: string;
  control_type: string;
  bounding_rect?: RawBoundingRect | null;
}

type RawAutomationQuery = Record<string, unknown>;

function normalizeAutomationElement(raw: RawAutomationElementInfo): AutomationElementInfo {
  return {
    id: raw.id,
    name: raw.name,
    className: raw.class_name,
    controlType: raw.control_type,
    boundingRect: raw.bounding_rect
      ? {
          left: raw.bounding_rect.left,
          top: raw.bounding_rect.top,
          width: raw.bounding_rect.width,
          height: raw.bounding_rect.height,
        }
      : null,
  };
}

function buildAutomationQuery(query: AutomationQuery): RawAutomationQuery {
  const payload: RawAutomationQuery = {};
  if (query.parentId) payload['parent_id'] = query.parentId;
  if (query.window) payload['window'] = query.window;
  if (query.windowClass) payload['window_class'] = query.windowClass;
  if (query.name) payload['name'] = query.name;
  if (query.className) payload['class_name'] = query.className;
  if (query.automationId) payload['automation_id'] = query.automationId;
  if (query.controlType) payload['control_type'] = query.controlType;
  if (query.maxResults != null) payload['max_results'] = query.maxResults;
  return payload;
}

function buildClickRequest(request: AutomationClickRequest): RawAutomationQuery {
  const payload: RawAutomationQuery = {};
  if (request.elementId) payload['element_id'] = request.elementId;
  if (request.x != null) payload['x'] = request.x;
  if (request.y != null) payload['y'] = request.y;
  if (request.button) payload['button'] = request.button;
  return payload;
}

function buildScreenshotRequest(options: AutomationScreenshotOptions): RawAutomationQuery {
  const payload: RawAutomationQuery = {};
  if (options.elementId) payload['element_id'] = options.elementId;
  if (options.x != null) payload['x'] = options.x;
  if (options.y != null) payload['y'] = options.y;
  if (options.width != null) payload['width'] = options.width;
  if (options.height != null) payload['height'] = options.height;
  if (options.conversationId != null) payload['conversation_id'] = options.conversationId;
  return payload;
}

// Updated Nov 16, 2025: Added error handling and timeout
export async function listAutomationWindows(): Promise<AutomationElementInfo[]> {
  try {
    const raw = await invokeWithTimeout<RawAutomationElementInfo[]>('automation_list_windows');
    return raw.map(normalizeAutomationElement);
  } catch (error) {
    throw new Error(`Failed to list automation windows: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling, validation, and timeout
export async function findAutomationElements(
  query: AutomationQuery,
): Promise<AutomationElementInfo[]> {
  try {
    if (!query || Object.keys(query).length === 0) {
      throw new Error('Query object cannot be empty');
    }
    const raw = await invokeWithTimeout<RawAutomationElementInfo[]>('automation_find_elements', {
      request: buildAutomationQuery(query),
    });
    return raw.map(normalizeAutomationElement);
  } catch (error) {
    throw new Error(`Failed to find automation elements: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling, validation, and timeout
export async function invokeElement(elementId: string): Promise<void> {
  try {
    validateNonEmpty(elementId, 'elementId');
    await invokeWithTimeout('automation_invoke', { request: { element_id: elementId } });
  } catch (error) {
    throw new Error(`Failed to invoke element ${elementId}: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling, validation, and timeout
export async function setElementValue(
  elementId: string,
  value: string,
  focus = false,
): Promise<void> {
  try {
    validateNonEmpty(elementId, 'elementId');
    if (value === undefined || value === null) {
      throw new Error('value cannot be null or undefined');
    }
    await invokeWithTimeout('automation_set_value', {
      request: {
        element_id: elementId,
        value,
        focus,
      },
    });
  } catch (error) {
    throw new Error(`Failed to set element value for ${elementId}: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling, validation, and timeout
export async function getElementValue(elementId: string): Promise<string> {
  try {
    validateNonEmpty(elementId, 'elementId');
    return await invokeWithTimeout<string>('automation_get_value', { element_id: elementId });
  } catch (error) {
    throw new Error(`Failed to get element value for ${elementId}: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling, validation, and timeout
export async function toggleElement(elementId: string): Promise<void> {
  try {
    validateNonEmpty(elementId, 'elementId');
    await invokeWithTimeout('automation_toggle', { element_id: elementId });
  } catch (error) {
    throw new Error(`Failed to toggle element ${elementId}: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling, validation, and timeout
export async function focusWindow(elementId: string): Promise<void> {
  try {
    validateNonEmpty(elementId, 'elementId');
    await invokeWithTimeout('automation_focus_window', { element_id: elementId });
  } catch (error) {
    throw new Error(`Failed to focus window ${elementId}: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling, validation, and timeout
export async function sendKeys(
  text: string,
  options: { elementId?: string; x?: number; y?: number; focus?: boolean } = {},
): Promise<void> {
  try {
    if (text === undefined || text === null) {
      throw new Error('text cannot be null or undefined');
    }
    validateCoordinates(options.x, options.y);
    await invokeWithTimeout('automation_send_keys', {
      request: {
        text,
        element_id: options.elementId,
        x: options.x,
        y: options.y,
        focus: options.focus,
      },
    });
  } catch (error) {
    throw new Error(`Failed to send keys: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function sendHotkey(key: number, modifiers: string[]): Promise<void> {
  try {
    if (!Number.isInteger(key) || key < 0) {
      throw new Error(`Invalid key code: ${key}`);
    }
    if (!Array.isArray(modifiers)) {
      throw new Error('modifiers must be an array');
    }
    await invokeWithTimeout('automation_hotkey', { request: { key, modifiers } });
  } catch (error) {
    throw new Error(`Failed to send hotkey: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function clickAutomation(request: AutomationClickRequest): Promise<void> {
  try {
    if (!request || Object.keys(request).length === 0) {
      throw new Error('Click request cannot be empty');
    }
    validateCoordinates(request.x, request.y);
    await invokeWithTimeout('automation_click', { request: buildClickRequest(request) });
  } catch (error) {
    throw new Error(`Failed to perform automation click: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and timeout
export async function automationScreenshot(
  options: AutomationScreenshotOptions = {},
): Promise<CaptureResult> {
  try {
    validateCoordinates(options.x, options.y);
    if (options.width !== undefined && (options.width <= 0 || !Number.isFinite(options.width))) {
      throw new Error(`Invalid width: ${options.width}`);
    }
    if (options.height !== undefined && (options.height <= 0 || !Number.isFinite(options.height))) {
      throw new Error(`Invalid height: ${options.height}`);
    }
    const raw = await invokeWithTimeout<RawCaptureResult>(
      'automation_screenshot',
      { request: buildScreenshotRequest(options) },
      60000, // 60 seconds for screenshots
    );
    return normalizeCaptureResult(raw);
  } catch (error) {
    throw new Error(`Failed to capture automation screenshot: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function automationOcr(imagePath: string): Promise<AutomationOcrResult> {
  try {
    validateNonEmpty(imagePath, 'imagePath');
    return await invokeWithTimeout<AutomationOcrResult>(
      'automation_ocr',
      { image_path: imagePath },
      60000, // 60 seconds for OCR
    );
  } catch (error) {
    throw new Error(`Failed to perform OCR on ${imagePath}: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and timeout
export async function getClipboardText(): Promise<string> {
  try {
    return await invokeWithTimeout<string>('automation_clipboard_get');
  } catch (error) {
    throw new Error(`Failed to get clipboard text: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function setClipboardText(text: string): Promise<void> {
  try {
    if (text === undefined || text === null) {
      throw new Error('text cannot be null or undefined');
    }
    await invokeWithTimeout('automation_clipboard_set', { text });
  } catch (error) {
    throw new Error(`Failed to set clipboard text: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function emitOverlayClick(payload: OverlayClickPayload): Promise<void> {
  try {
    if (!payload) {
      throw new Error('payload cannot be null or undefined');
    }
    await invokeWithTimeout('overlay_emit_click', { payload });
  } catch (error) {
    throw new Error(`Failed to emit overlay click: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function emitOverlayType(payload: OverlayTypePayload): Promise<void> {
  try {
    if (!payload) {
      throw new Error('payload cannot be null or undefined');
    }
    await invokeWithTimeout('overlay_emit_type', { payload });
  } catch (error) {
    throw new Error(`Failed to emit overlay type: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function emitOverlayRegion(payload: OverlayRegionPayload): Promise<void> {
  try {
    if (!payload) {
      throw new Error('payload cannot be null or undefined');
    }
    await invokeWithTimeout('overlay_emit_region', { payload });
  } catch (error) {
    throw new Error(`Failed to emit overlay region: ${error}`);
  }
}

// Updated Nov 16, 2025: Added error handling and validation
export async function replayOverlayEvents(limit?: number): Promise<void> {
  try {
    if (limit !== undefined && (!Number.isInteger(limit) || limit <= 0)) {
      throw new Error(`Invalid limit: ${limit}`);
    }
    await invokeWithTimeout('overlay_replay_recent', { limit });
  } catch (error) {
    throw new Error(`Failed to replay overlay events: ${error}`);
  }
}

File: apps/desktop/src-tauri/src/commands/automation.rs
use anyhow::{anyhow, Result as AnyResult};
use serde::Deserialize;
use tauri::{AppHandle, Emitter, State};

use super::capture::{capture_screen_full, capture_screen_region};
use super::AppDatabase;
use crate::automation::screen::{perform_ocr, OcrResult};
use crate::{
    automation::{
        global_service,
        input::{KeyboardSimulator, MouseButton},
        uia::{ElementQuery, UIElementInfo},
        AutomationService,
    },
    db::{
        models::{OverlayEvent, OverlayEventType},
        repository,
    },
    overlay::{
        dispatch_overlay_animation, dispatch_overlay_animation_normalized, ensure_overlay_ready,
        OverlayAnimation,
    },
};

#[derive(Debug, Deserialize)]
pub struct FindElementsRequest {
    #[serde(default)]
    pub parent_id: Option<String>,
    #[serde(default)]
    pub window: Option<String>,
    #[serde(default)]
    pub window_class: Option<String>,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub class_name: Option<String>,
    #[serde(default)]
    pub automation_id: Option<String>,
    #[serde(default)]
    pub control_type: Option<String>,
    #[serde(default)]
    pub max_results: Option<usize>,
}

#[derive(Debug, Deserialize)]
pub struct InvokeRequest {
    pub element_id: String,
}

#[derive(Debug, Deserialize)]
pub struct ValueRequest {
    pub element_id: String,
    pub value: String,
    #[serde(default)]
    pub focus: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct ClickRequest {
    #[serde(default)]
    pub element_id: Option<String>,
    #[serde(default)]
    pub x: Option<i32>,
    #[serde(default)]
    pub y: Option<i32>,
    #[serde(default)]
    pub button: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ScreenshotRequest {
    #[serde(default)]
    pub element_id: Option<String>,
    #[serde(default)]
    pub x: Option<i32>,
    #[serde(default)]
    pub y: Option<i32>,
    #[serde(default)]
    pub width: Option<u32>,
    #[serde(default)]
    pub height: Option<u32>,
    #[serde(default)]
    pub conversation_id: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct OverlayClickPayload {
    pub x: i32,
    pub y: i32,
    #[serde(default = "default_left_button")]
    pub button: String,
}

#[derive(Debug, Deserialize)]
pub struct OverlayTypePayload {
    pub x: i32,
    pub y: i32,
    pub text: String,
}

#[derive(Debug, Deserialize)]
pub struct OverlayRegionPayload {
    pub x: i32,
    pub y: i32,
    pub width: i32,
    pub height: i32,
}

#[derive(Debug, Deserialize)]
pub struct SendKeysRequest {
    pub text: String,
    #[serde(default)]
    pub element_id: Option<String>,
    #[serde(default)]
    pub x: Option<i32>,
    #[serde(default)]
    pub y: Option<i32>,
    #[serde(default)]
    pub focus: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct HotkeyRequest {
    pub key: u16,
    #[serde(default)]
    pub modifiers: Vec<String>,
}

#[derive(Debug, Deserialize)]
pub struct DragDropRequest {
    pub from_x: i32,
    pub from_y: i32,
    pub to_x: i32,
    pub to_y: i32,
    #[serde(default = "default_drag_duration")]
    pub duration_ms: u32,
}

fn default_drag_duration() -> u32 {
    300
}

#[tauri::command]
pub fn automation_list_windows(app: AppHandle) -> Result<Vec<UIElementInfo>, String> {
    ensure_overlay_ready(&app);
    with_service(|service| service.uia.list_windows()).map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_find_elements(
    request: FindElementsRequest,
) -> Result<Vec<UIElementInfo>, String> {
    let query = ElementQuery {
        window: request.window,
        window_class: request.window_class,
        name: request.name,
        class_name: request.class_name,
        automation_id: request.automation_id,
        control_type: request.control_type,
        max_results: request.max_results,
    };

    with_service(|service| service.uia.find_elements(request.parent_id, &query))
        .map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_invoke(request: InvokeRequest) -> Result<(), String> {
    with_service(|service| service.uia.invoke(&request.element_id)).map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_set_value(request: ValueRequest) -> Result<(), String> {
    with_service(|service| {
        if request.focus.unwrap_or(false) {
            service.uia.set_focus(&request.element_id)?;
        }
        service.uia.set_value(&request.element_id, &request.value)
    })
    .map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_get_value(element_id: String) -> Result<String, String> {
    with_service(|service| service.uia.get_value(&element_id)).map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_get_text(element_id: String) -> Result<String, String> {
    automation_get_value(element_id)
}

#[tauri::command]
pub fn automation_toggle(element_id: String) -> Result<(), String> {
    with_service(|service| service.uia.toggle(&element_id)).map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_focus_window(element_id: String) -> Result<(), String> {
    with_service(|service| service.uia.focus_window(&element_id)).map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn automation_send_keys(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    request: SendKeysRequest,
) -> Result<(), String> {
    execute_text_input(&app, &db, &request, false).await
}

#[tauri::command]
pub fn automation_hotkey(request: HotkeyRequest) -> Result<(), String> {
    let modifiers: Vec<u16> = request
        .modifiers
        .iter()
        .filter_map(|name| KeyboardSimulator::modifier_key(name))
        .collect();

    with_service(|service| service.keyboard.hotkey(&modifiers, request.key))
        .map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_click(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    request: ClickRequest,
) -> Result<(), String> {
    ensure_overlay_ready(&app);
    let (x, y, button_name) = with_service(|service| {
        let (x, y) = if let Some(element_id) = &request.element_id {
            let rect = service
                .uia
                .bounding_rect(element_id)?
                .ok_or_else(|| anyhow!("Element {element_id} has no bounding rectangle"))?;
            let x = (rect.left + rect.width / 2.0).round() as i32;
            let y = (rect.top + rect.height / 2.0).round() as i32;
            (x, y)
        } else if let (Some(x), Some(y)) = (request.x, request.y) {
            (x, y)
        } else {
            return Err(anyhow!("Either element_id or (x, y) must be provided"));
        };

        let button_name = request.button.as_deref().unwrap_or("left").to_lowercase();
        let button = match button_name.as_str() {
            "right" => MouseButton::Right,
            "middle" => MouseButton::Middle,
            _ => MouseButton::Left,
        };

        service.mouse.click(x, y, button)?;
        Ok((x, y, button_name))
    })
    .map_err(|err| err.to_string())?;

    if let Ok(conn) = db.0.lock() {
        if let Err(err) = dispatch_overlay_animation(
            &app,
            &conn,
            OverlayAnimation::Click {
                x,
                y,
                button: button_name,
            },
        ) {
            tracing::warn!("Failed to emit click overlay animation: {err:?}");
        }
    }

    Ok(())
}

#[tauri::command]
pub async fn automation_type(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    request: SendKeysRequest,
) -> Result<(), String> {
    execute_text_input(&app, &db, &request, true).await
}

#[tauri::command]
pub async fn automation_drag_drop(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    request: DragDropRequest,
) -> Result<(), String> {
    ensure_overlay_ready(&app);

    // Create mouse simulator outside the service to avoid async closure issues
    let mouse = crate::automation::input::MouseSimulator::new().map_err(|e| e.to_string())?;
    mouse
        .drag_and_drop(
            request.from_x,
            request.from_y,
            request.to_x,
            request.to_y,
            request.duration_ms,
        )
        .await
        .map_err(|err| err.to_string())?;

    // Emit overlay animation for drag-drop
    if let Ok(conn) = db.0.lock() {
        if let Err(err) = dispatch_overlay_animation(
            &app,
            &conn,
            OverlayAnimation::RegionHighlight {
                x: request.from_x.min(request.to_x),
                y: request.from_y.min(request.to_y),
                width: (request.to_x - request.from_x).abs(),
                height: (request.to_y - request.from_y).abs(),
            },
        ) {
            tracing::warn!("Failed to emit drag-drop overlay animation: {err:?}");
        }
    }

    Ok(())
}

#[tauri::command]
pub fn automation_clipboard_get() -> Result<String, String> {
    with_service(|service| service.clipboard.get_text()).map_err(|err| err.to_string())
}

#[tauri::command]
pub fn automation_clipboard_set(text: String) -> Result<(), String> {
    with_service(|service| service.clipboard.set_text(&text)).map_err(|err| err.to_string())
}

#[tauri::command]
pub async fn automation_ocr(image_path: String) -> Result<OcrResult, String> {
    #[cfg(feature = "ocr")]
    {
        perform_ocr(&image_path)
            .await
            .map_err(|err| err.to_string())
    }
    #[cfg(not(feature = "ocr"))]
    {
        perform_ocr(&image_path).map_err(|err| err.to_string())
    }
}

#[tauri::command]
pub async fn automation_screenshot(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    request: ScreenshotRequest,
) -> Result<crate::commands::capture::CaptureResult, String> {
    ensure_overlay_ready(&app);

    if let Some(ref element_id) = request.element_id {
        let bounds = with_service(|service| service.uia.bounding_rect(element_id))
            .map_err(|err| err.to_string())?;
        if let Some(bounds) = bounds {
            let width = bounds.width.round().max(1.0) as u32;
            let height = bounds.height.round().max(1.0) as u32;
            let x = bounds.left.round() as i32;
            let y = bounds.top.round() as i32;
            return capture_screen_region(app, db, x, y, width, height, request.conversation_id)
                .await;
        }
    }

    if let (Some(x), Some(y), Some(width), Some(height)) =
        (request.x, request.y, request.width, request.height)
    {
        return capture_screen_region(app, db, x, y, width, height, request.conversation_id).await;
    }

    capture_screen_full(app, db, request.conversation_id).await
}

fn with_service<F, T>(operation: F) -> AnyResult<T>
where
    F: FnOnce(&mut AutomationService) -> AnyResult<T>,
{
    let mut guard = global_service()?;
    let service = guard
        .as_mut()
        .ok_or_else(|| anyhow!("Automation service unavailable"))?;
    operation(service)
}

async fn execute_text_input(
    app: &AppHandle,
    db: &State<'_, AppDatabase>,
    request: &SendKeysRequest,
    force_focus: bool,
) -> Result<(), String> {
    ensure_overlay_ready(app);

    let text = request.text.clone();
    let element_id = request.element_id.clone();
    let fallback_position = request.x.zip(request.y);
    let should_focus = force_focus || request.focus.unwrap_or(false);

    // Create a keyboard simulator outside the closure
    let keyboard = KeyboardSimulator::new().map_err(|e| e.to_string())?;

    let location = with_service(|service| {
        if let Some(element_id) = &element_id {
            if should_focus {
                let _ = service.uia.set_focus(element_id);
            }

            if let Some(bounds) = service.uia.bounding_rect(element_id)? {
                let x = (bounds.left + bounds.width / 2.0).round() as i32;
                let y = (bounds.top + bounds.height / 2.0).round() as i32;
                return Ok(Some((x, y)));
            }
        }

        Ok(fallback_position)
    })
    .map_err(|err| err.to_string())?;

    // Send text asynchronously outside the closure
    keyboard.send_text(&text).await.map_err(|e| e.to_string())?;

    if let Ok(conn) = db.0.lock() {
        if let Err(err) = dispatch_overlay_animation(
            app,
            &conn,
            OverlayAnimation::Type {
                x: location.map(|(x, _)| x).unwrap_or(0),
                y: location.map(|(_, y)| y).unwrap_or(0),
                text: text.chars().take(32).collect(),
            },
        ) {
            tracing::warn!("Failed to emit typing overlay animation: {err:?}");
        }
    }

    Ok(())
}

#[tauri::command]
pub fn overlay_emit_click(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    payload: OverlayClickPayload,
) -> Result<(), String> {
    ensure_overlay_ready(&app);
    if let Ok(conn) = db.0.lock() {
        dispatch_overlay_animation_normalized(
            &app,
            &conn,
            OverlayAnimation::Click {
                x: payload.x,
                y: payload.y,
                button: payload.button,
            },
        )
        .map_err(|err| err.to_string())?;
    }
    Ok(())
}

#[tauri::command]
pub fn overlay_emit_type(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    payload: OverlayTypePayload,
) -> Result<(), String> {
    ensure_overlay_ready(&app);
    if let Ok(conn) = db.0.lock() {
        dispatch_overlay_animation_normalized(
            &app,
            &conn,
            OverlayAnimation::Type {
                x: payload.x,
                y: payload.y,
                text: payload.text,
            },
        )
        .map_err(|err| err.to_string())?;
    }
    Ok(())
}

#[tauri::command]
pub fn overlay_emit_region(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    payload: OverlayRegionPayload,
) -> Result<(), String> {
    ensure_overlay_ready(&app);
    if let Ok(conn) = db.0.lock() {
        dispatch_overlay_animation_normalized(
            &app,
            &conn,
            OverlayAnimation::RegionHighlight {
                x: payload.x,
                y: payload.y,
                width: payload.width,
                height: payload.height,
            },
        )
        .map_err(|err| err.to_string())?;
    }
    Ok(())
}

#[tauri::command]
pub fn overlay_replay_recent(
    app: AppHandle,
    db: State<'_, AppDatabase>,
    limit: Option<usize>,
) -> Result<(), String> {
    ensure_overlay_ready(&app);
    let events = {
        let conn = db.0.lock().map_err(|e| e.to_string())?;
        repository::list_overlay_events(&conn, None, None).map_err(|e| e.to_string())?
    };

    let limit = limit.unwrap_or(10);
    let count = events.len();
    let start = count.saturating_sub(limit);

    for event in events.into_iter().skip(start) {
        if let Some(animation) = animation_from_event(event) {
            emit_overlay(&app, &animation);
        }
    }

    Ok(())
}

fn default_left_button() -> String {
    "left".to_string()
}

fn emit_overlay(app: &AppHandle, animation: &OverlayAnimation) {
    let _ = app.emit(animation.event_name(), animation);
    let _ = app.emit("overlay://event", animation);
}

fn animation_from_event(event: OverlayEvent) -> Option<OverlayAnimation> {
    match event.event_type {
        OverlayEventType::ScreenshotFlash => Some(OverlayAnimation::ScreenshotFlash),
        _ => event
            .data
            .as_deref()
            .and_then(|json| serde_json::from_str::<OverlayAnimation>(json).ok()),
    }
}
